### PR TITLE
msys2: use sourceforge urls

### LINKED ISSUE #1055: [package] msys2/20190524: URL to the msys repo is dead
### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **msys2/20161025**
  * Operating System+version: **Windows 10**
  * Compiler+version: **Visual Studio 16**
  * Conan version: **conan 1.22.2**
  * Python version: **Python 3.8.1**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
[settings]
os=Windows
os_build=Windows
arch=x86_64
arch_build=x86_64
compiler=Visual Studio
compiler.version=16
build_type=Release
[options]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)
I tried to build a QT package, which has an indirect dependency to the msys2 package.
When conan tries to build the package, throws an HTTP connection exception. (See the logs)
Opening the url in the conandata.yml could not be opened either.

### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```
msys2/20161025: Applying build-requirement: 7zip/19.00
msys2/20161025: WARN: Build folder is dirty, removing it: C:\.conan\ddad51\1
msys2/20161025: Copying sources to build folder
msys2/20161025: Building your package in C:\.conan\ddad51\1
msys2/20161025: Generator txt created conanbuildinfo.txt
msys2/20161025: Calling build()
ERROR: Error downloading file http://repo.msys2.org/distrib/x86_64/msys2-base-x86_64-20161025.tar.xz: 'HTTPConnectionPool(host='repo.msys2.org', port=80): Max retries exceeded with url: /distrib/x86_64/msys2-base-x86_64-20161025.tar.xz (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x04480820>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'))'
Waiting 5 seconds to retry...
msys2/20161025:
msys2/20161025: ERROR: Package '3cbd862cb7a3ea47ed1220dc027f344262d5224f' build failed
msys2/20161025: WARN: Build folder C:\.conan\ddad51\1
ERROR: msys2/20161025: Error in build() method, line 48
        filename = self._download(**self.conan_data["sources"][self.version][arch])
while calling '_download', line 36
        tools.download(url, filename)
        ConanException: Error downloading file http://repo.msys2.org/distrib/x86_64/msys2-base-x86_64-20161025.tar.xz: 'HTTPConnectionPool(host='repo.msys2.org', port=80): Max retries exceeded with url: /distrib/x86_64/msys2-base-x86_64-20161025.tar.xz (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x04480D00>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'))'
```

</details>

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **msys2/{20190524,20061025}**

Fixes #1055

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Croydon** [comment]
IIRC then using `https://downloads.sourceforge.net...` URLs is better. As the other download urls have a waiting period which the download subdomain doesn't have

**Comment by SSE4** [comment]
what's the advantage of using source forge? AFAIK their server often experience major outages (that's why we moved bzip2 to bintray I guess?)

**Comment by madebr** [comment]
@SSE4 
I used sourceforge as that source is well known.
But if what you're saying is true, we might want to consider adding support for alternatives.
Temporary unavailabilities occur all the time.
The openssl recipe is doing something similar:

https://github.com/conan-io/conan-center-index/blob/6a86b1c0264e6771af0cdd6ec309833020bb8b4e/recipes/openssl/ALL/conanfile.py#L133-L139

What would be nice if we could specify a list of alternatives in `conandata.yml` and `tools.get` would just try all in the list.

Something like this:
```
sources:
  "20161025":
    - sha256: "8bafd3d52f5a51528a8671c1cae5591b36086d6ea5b1e76e17e390965cf6768f"
      url:
      - "http://repo.msys2.org/distrib/i686/msys2-base-i686-20161025.tar.xz"
      - "https://sourceforge.net/projects/msys2/files/Base/i686/msys2-base-i686-20161025.tar.xz"
    - sha256: "bb1f1a0b35b3d96bf9c15092da8ce969a84a134f7b08811292fbc9d84d48c65d"
      url:
      - "http://repo.msys2.org/distrib/x86_64/msys2-base-x86_64-20161025.tar.xz"
      - "https://sourceforge.net/projects/msys2/files/Base/x86_64/msys2-base-x86_64-20161025.tar.xz"
```
To avoid doing this again and again for other recipes.

**Comment by SSE4** [comment]
@madebr OpenSSL is a bit different story. they host only latest release(s) on [https://www.openssl.org/source/](https://www.openssl.org/source/), but as soon as new release is out, they move previous one to the [https://www.openssl.org/source/old/](https://www.openssl.org/source/old/)
but I agree, we can use mirror(s) as a fallback

**Comment by madebr** [comment]
@uilianries Does Conan have some built-in functionality that does this? What do you think of my proposal?

**Comment by uilianries** [comment]
@madebr I remember that one recipe has an alternative url for download. But built-in function, unfortunately not.

**Comment by uilianries** [comment]
For mirrors, we can use Bintray. I remember that Boost is mirrored there.

**Comment by madebr** [comment]
> For mirrors, we can use Bintray. I remember that Boost is mirrored there.

If neither msys2 or sourceforge have a large enough uptime, then bintray would indeed be a nice mirror.
But I'm not the one making that decision...

**Comment by danimtb** [comment]
The original urls are working ok now and I think it will be better to have the mirror feature than replacing URLs every now and then. It will be available soon in next Conan release https://github.com/conan-io/conan/pull/6679

**Comment by madebr** [comment]
@danimtb Instead of closing, wouldn't it be better to tag these issues/prs with `conan-release-1.24` or `conan-next-release`?

**Comment by danimtb** [comment]
Sorry about that. As the issue is still open I thought it was better to continue the conversation there and create a new PR for it. Feel free to reopen if you wish

**Comment by madebr** [comment]
No problem mate. I don't think I have sufficient rights to re-open closed issues.

**Comment by jgsogo** [comment]
Conan 1.24 has support for mirrors (https://docs.conan.io/en/latest/reference/tools.html?highlight=mirrors). It should be enough to change the `yml` and provide a list to the `url` key.

Will the hooks validate this YAML structure? 🙈

**Comment by uilianries** [comment]
> Will the hooks validate this YAML structure?

No, actually it validates the keywords only. List are allowed by default.

**Comment by madebr** [comment]
@danimtb @uilianries @SSE4 Please re-review

**Comment by jgsogo** [comment]
The CI reports an error trying to remove the folder where this job is running 🤔 

```
 Unable to delete 'C:\J\c3ipr\pr_1060_8_2_0'
```

**Comment by madebr** [comment]
@jgsogo 
Indeed, the build was successful at https://c3i.jfrog.io/c3i/misc/logs/pr/1060/8/msys2/20190524/
There is probably a problem with removing some *special* files such as `/dev/stdout`.

**Comment by jgsogo** [comment]
Yes, we'll investigate on the CI side, and we'll use this PR to play a little bit with this error.

**Comment by jgsogo** [comment]
I'm having a look at this one, I get the same error in our `staging` infrastructure. There are some errors saying `Unable to remove file ....` for `exe` files and `dll`. Strange 🧐

**Comment by jgsogo** [comment]
The failure is related to the `dirmngr.exe` executable. After packaging the Conan package, the workspace is cleaned up, but the executable is still running and fails to remove it (and some shared libraries)

image

It is the `dirmngr.exe` in the build directory.

**Comment by jgsogo** [comment]
@madebr , is this happening to you locally?

**Comment by madebr** [comment]
> @madebr , is this happening to you locally?

That process is also running on my pc after the `build()` step.
I've added a patch that tries to kill all processes, running in the build folder.

I'm curious whether this will fix anything on the ci.

**Comment by madebr** [comment]
This *worked on my computer*:tm:.
I could remove the build folder after that without issues.
But this will probably be more complicated on a server.

<details><summary>Output of the _kill_processes_in_build_folder method on my pc</summary>
<p>

```
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying C:\WINDOWS\system32\sihost.exe
trying C:\WINDOWS\system32\svchost.exe
trying
trying C:\WINDOWS\system32\svchost.exe
trying
trying C:\WINDOWS\system32\taskhostw.exe
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying
trying C:\Program Files\Elantech\ETDCtrl.exe
trying
trying
trying
trying
trying
trying C:\WINDOWS\system32\igfxEM.exe
trying C:\WINDOWS\system32\igfxHK.exe
trying C:\WINDOWS\system32\igfxTray.exe
trying
trying
trying C:\Program Files\NVIDIA Corporation\Display\nvtray.exe
trying
trying C:\WINDOWS\Explorer.EXE
trying
trying
trying
trying C:\WINDOWS\system32\svchost.exe
trying
trying
trying C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy\StartMenuExperienceHost.exe
trying C:\Windows\System32\RuntimeBroker.exe
trying C:\Windows\SystemApps\Microsoft.Windows.Search_cw5n1h2txyewy\SearchApp.exe
trying
trying C:\Windows\System32\RuntimeBroker.exe
trying
trying C:\Windows\SystemApps\Microsoft.LockApp_cw5n1h2txyewy\LockApp.exe
trying C:\Windows\System32\RuntimeBroker.exe
trying C:\Windows\System32\RuntimeBroker.exe
trying
trying C:\Program Files\WindowsApps\Microsoft.YourPhone_1.20041.91.0_x64__8wekyb3d8bbwe\YourPhone.exe
trying
trying C:\Windows\System32\RuntimeBroker.exe
trying C:\Program Files\NVIDIA Corporation\Display\nvtray.exe
trying C:\Windows\System32\SecurityHealthSystray.exe
trying
trying C:\Program Files (x86)\NVIDIA Corporation\Update Core\NvBackend.exe
trying C:\Program Files\Realtek\Audio\HDA\RAVCpl64.exe
trying C:\Program Files\Realtek\Audio\HDA\RAVBg64.exe
trying C:\Windows\System32\RuntimeBroker.exe
trying C:\WINDOWS\system32\ApplicationFrameHost.exe
trying C:\Windows\SystemApps\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\MicrosoftEdge.exe
trying C:\WINDOWS\system32\browser_broker.exe
trying C:\Windows\System32\RuntimeBroker.exe
trying C:\WINDOWS\system32\MicrosoftEdgeSH.exe
trying C:\Windows\System32\MicrosoftEdgeCP.exe
trying C:\Program Files\WindowsApps\Microsoft.WindowsStore_12005.1001.1.0_x64__8wekyb3d8bbwe\WinStore.App.exe
trying
trying C:\Windows\SystemApps\MicrosoftWindows.Client.CBS_cw5n1h2txyewy\InputApp\TextInputHost.exe
trying C:\WINDOWS\system32\DllHost.exe
trying C:\Windows\ImmersiveControlPanel\SystemSettings.exe
trying C:\WINDOWS\system32\cmd.exe
trying C:\WINDOWS\system32\conhost.exe
trying C:\WINDOWS\system32\svchost.exe
trying
trying
trying
trying
trying
trying
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying
trying
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\WINDOWS\System32\Wbem\WMIC.exe
trying C:\Program Files (x86)\JetBrains\PyCharm Community Edition 2020.1.1\bin\pycharm64.exe
trying C:\Program Files (x86)\JetBrains\PyCharm Community Edition 2020.1.1\bin\fsnotifier64.exe
trying C:\WINDOWS\system32\conhost.exe
trying C:\Program Files\WindowsApps\Microsoft.WindowsTerminal_1.0.1401.0_x64__8wekyb3d8bbwe\WindowsTerminal.exe
trying C:\Program Files\WindowsApps\Microsoft.WindowsTerminal_1.0.1401.0_x64__8wekyb3d8bbwe\OpenConsole.exe
trying C:\WINDOWS\system32\cmd.exe
trying C:\msys64\usr\bin\bash.exe
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\Windows\System32\CompPkgSrv.exe
trying C:\Program Files\WindowsApps\Microsoft.WindowsTerminal_1.0.1401.0_x64__8wekyb3d8bbwe\OpenConsole.exe
trying C:\WINDOWS\system32\cmd.exe
trying C:\Program Files\WindowsApps\Microsoft.Windows.Photos_2020.19111.24110.0_x64__8wekyb3d8bbwe\Microsoft.Photos.exe
trying
trying C:\Windows\System32\RuntimeBroker.exe
trying C:\Program Files\Git\mingw64\bin\wish.exe
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\MSBuild.exe
trying C:\WINDOWS\system32\conhost.exe
trying C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\MSBuild.exe
trying C:\WINDOWS\system32\conhost.exe
trying C:\Python38\Scripts\conan.exe
trying c:\python38\python.exe
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\WINDOWS\system32\conhost.exe
trying C:\.conan\a9447c\1\msys64\usr\bin\gpg-agent.exe
msys2/20190524: Name: gpg-agent.exe
msys2/20190524: ExecutablePath: C:\.conan\a9447c\1\msys64\usr\bin\gpg-agent.exe
msys2/20190524: ProcessId: 8572
msys2/20190524: Attempting to kill 8572

----Running------
> wmic process where ProcessId=8572 delete
-----------------
Deleting instance \\MAARTEN-ACER2\ROOT\CIMV2:Win32_Process.Handle="8572"
Instance deletion successful.
trying C:\WINDOWS\system32\conhost.exe
trying C:\.conan\a9447c\1\msys64\usr\bin\dirmngr.exe
msys2/20190524: Name: dirmngr.exe
msys2/20190524: ExecutablePath: C:\.conan\a9447c\1\msys64\usr\bin\dirmngr.exe
msys2/20190524: ProcessId: 7864
msys2/20190524: Attempting to kill 7864

----Running------
> wmic process where ProcessId=7864 delete
-----------------
Deleting instance \\MAARTEN-ACER2\ROOT\CIMV2:Win32_Process.Handle="7864"
Instance deletion successful.
trying
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying
trying C:\Program Files\Mozilla Firefox\firefox.exe
trying C:\WINDOWS\system32\cmd.exe
trying C:\WINDOWS\System32\Wbem\WMIC.exe
trying
```

</p>
</details>

**Comment by jgsogo** [comment]
I think this is something we need to fix in the infra, that function doesn't belong to a Conan recipe. 

There should be a way to kill all the processes spawned from a CI job 🧑‍🏭

**Comment by madebr** [comment]
@jgsogo 
After the CI problem gets solved, I'd love to hear how you did it 😄

**Comment by jgsogo** [comment]
I've asked the people taking care of the infrastructure, they might know more than me about CI in general. We have a workaround: do not remove the workspace after the job and, eventually, some cleanup task will take care of it (and after restarting the machine the process will no longer be there).

**Comment by jgsogo** [comment]
😊  it works!

Please, remove the `_kill_processes_in_build_folder` function as it is no longer needed.

Thanks!

File: recipes/msys2/all/conanfile.py [hunk 1]
```diff
@@ -30,11 +32,36 @@ def source(self):
...
+            if self.build_folder.lower() in row["ExecutablePath"].lower() or \
+                    self.build_folder.replace("\\", "/").lower() in row["ExecutablePath"].lower() or \
+                    tools.unix_path(self.build_folder).lower() in row["ExecutablePath"].lower():
+                self.output.info("Name: " + row["Name"])
+                self.output.info("ExecutablePath: " +  row["ExecutablePath"])
+                self.output.info("ProcessId: " +  row["ProcessId"])
+                self.output.info("Attempting to kill " + row["ProcessId"])
+                self.run("wmic process where ProcessId={} delete".format(row["ProcessId"]))
+
     def _download(self, url, sha256):
```
Comment (madebr) [inline recipes/msys2/all/conanfile.py:61]:
```suggestion
        pass

    def _download(self, url, sha256):
```

File: recipes/msys2/all/conanfile.py [hunk 2]
```diff
@@ -75,6 +100,8 @@ def build(self):
         tools.replace_in_file(os.path.join(self._msys_dir, "etc", "profile"),
                               'PKG_CONFIG_PATH="', 'PKG_CONFIG_PATH="$PKG_CONFIG_PATH:')
 
+        self._kill_processes_in_build_folder()
+
     def package(self):
```
Comment (madebr) [inline recipes/msys2/all/conanfile.py:105]:
```suggestion
                              'PKG_CONFIG_PATH="', 'PKG_CONFIG_PATH="$PKG_CONFIG_PATH:')

    def package(self):
```

File: recipes/msys2/all/conanfile.py [hunk 3]
```diff
@@ -1,5 +1,7 @@
 from conans import ConanFile, tools
 from conans.errors import ConanInvalidConfiguration
+import csv
+from io import StringIO
```
Comment (danimtb) [inline recipes/msys2/all/conanfile.py:4]: Not used?

Comment (madebr) [inline recipes/msys2/all/conanfile.py:4]:
```suggestion
```

Comment (madebr) [reply recipes/msys2/all/conanfile.py]: Indeed!

**Comment by jgsogo** [comment]
The first message belongs to one of the previous commits, that was canceled with the following commits. We need to add to every message the commit it belongs to.

image

**Pull request merged**